clarity/src/vm/ast/parser_v2/lexer/mod.rs

```diff
@@ -0,0 +1,1957 @@
+
+        lexer = Lexer::new("{", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Lbrace);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("}", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Rbrace);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new(":", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Colon);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new(",", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Comma);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new(".", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Dot);
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("123", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Int(123));
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("0123", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Int(123));
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("0", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Int(0));
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("0a", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Int(0));
+        assert_eq!(lexer.diagnostics.len(), 1);
+        assert_eq!(lexer.diagnostics[0].e, LexerError::InvalidCharInt('a'));
+
+        lexer = Lexer::new("56789*", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Int(56789));
+        assert_eq!(lexer.diagnostics.len(), 1);
+        assert_eq!(lexer.diagnostics[0].e, LexerError::InvalidCharInt('*'));
+
+        lexer = Lexer::new("u123", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Uint(123));
+
+        lexer = Lexer::new("u1a", false);
+        assert_eq!(lexer.read_token().unwrap().token, Token::Uint(1));
+        assert_eq!(lexer.diagnostics.len(), 1);
+        assert_eq!(lexer.diagnostics[0].e, LexerError::InvalidCharUint('a'));
+
+        lexer = Lexer::new("\"hello\"", false);
+        assert_eq!(
+            lexer.read_token().unwrap().token,
+            Token::AsciiString("hello".to_string())
+        );
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("\"new\\nline\"", false);
+        assert_eq!(
+            lexer.read_token().unwrap().token,
+            Token::AsciiString("new\nline".to_string())
+        );
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("\"quote \\\"this\\\"\"", false);
+        assert_eq!(
+            lexer.read_token().unwrap().token,
+            Token::AsciiString("quote \"this\"".to_string())
+        );
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("\"\\r\\t\\0\\\\ ok\"", false);
+        assert_eq!(
```

Comment:
I don't think this is allowed today in Clarity -- you can't have bytes`\r` or `\0` in the code body at all.

Reply:
Inside a string, they are allowed in the current parser, though I am noticing now that the `\0` will later trigger an error (the `\r` is allowed) from:

https://github.com/stacks-network/stacks-blockchain/blob/38aa968d7e81d575453ba5253aa8e64a8ac7cec9/clarity/src/vm/types/mod.rs#L767-L771

So perhaps instead we should just disallow `\0` in the parser? And also remove `\r` if we don't want it to be supported.